(base) [malhal@g-12-l0002 final_csi_results]$ cat ecoli.pbs
#!/bin/bash
#PBS -N july_salmonella
#PBS -o july_salmonella.output
#PBS -e july_salmonella.error
#PBS -l walltime=128:00:00
#PBS -l select=1:ncpus=20:mem=128gb

# Load any modules or set up environment variables if needed
module load tools
module load anaconda3/2021.11
module load perl
module load bwa/0.7.12
module load bedtools/2.30.0
module load jdk/18.0.1
module load cromwell/72
module load FastTree/2.1.11
module load mummer/3.23
module load samtools/0.1.18

# Move to the directory where the job was submitted
cd $PBS_O_WORKDIR

# Your commands go here, for example, a Python script execution
source /home/projects/cge/apps/env/csi/bin/activate
/usr/bin/time -v  python /home/projects/cge/apps/CSIPhylogeny/csi_phylogeny.py -o  /home/projects/cge/people/malhal/final_csi_results/july_salmonella -t  /home/projects/cge/people/malhal/final_csi_results/tmp_july_salmonella/ -r /home/projects/cge/people/malhal/final_csi_results/salmonella.fasta --aligner_path /services/tools/bwa/0.7.12/bin/bwa --samtools_path /services/tools/samtools/0.1.18/bin/samtools --vcfutils_path /services/tools/samtools/0.1.18/bin/vcfutils.pl --genomeCoverageBed_path /services/tools/bedtools/2.30.0/bin/genomeCoverageBed --bcftools_path /services/tools/samtools/0.1.18/bin/bcftools --perl_path /services/tools/perl/5.20.1/bin/perl5.20.1 --nucmer_path /services/tools/mummer/3.23/nucmer --cromwell_path /services/tools/cromwell/72/cromwell-72.jar --cromwell_cfg /home/projects/cge/apps/CSIPhylogeny/cromwell_alt.cfg --delta_filter_path /services/tools/mummer/3.23/delta-filter --show_coords_path /services/tools/mummer/3.23/show-coords --show_snps_path /services/tools/mummer/3.23/show-snps --fasttree_path /services/tools/fasttree/2.1.11/FastTree /home/projects/cge/people/malhal/cgphylo/salmonella_enterica/*_1.fastq.gz